packages/tailwindcss/src/theme-fn.ts

```diff
@@ -0,0 +1,152 @@
+import { deepMerge } from './compat/config/deep-merge'
+import type { UserConfig } from './compat/config/types'
+import type { DesignSystem } from './design-system'
+import type { Theme } from './theme'
+import { DefaultMap } from './utils/default-map'
+import { toKeyPath } from './utils/to-key-path'
+
+export function createThemeFn(
+  designSystem: DesignSystem,
+  configTheme: () => UserConfig['theme'],
+  resolveValue: (value: any) => any,
+) {
+  return function theme(path: string) {
+    let keypath = toKeyPath(path)
+    let cssValue = readFromCss(designSystem.theme, keypath)
+
+    // If we get a single non-null value back from the CSS theme, return it
+    // Otherwise, we need to merge the CSS theme with the one resolved from
+    // all registered plugins.
+    if (cssValue === null || typeof cssValue !== 'object') {
+      return cssValue
+    }
```

Comment:
Noticed one bug while rebasing the CSS `theme()` function. When we read form the CSS, we now no longer replace `colors` in the namespace with `color`. which means that `theme(colors.red.500)` won't work correctly.

Reply:
yeep… I noticed this last night when working on some compatibility stuff. I think this means we might need to bring the explicit theme upgrade list back unfortunately. I'd prefer _not_ to if we can figure out a way though.

Reply:
This is even more annoying for things like `theme(backgroundColors.red.500)` where even if we replace `colors` to color`, it will still not be able to find the right CSS variable because we have no mapping to the CSS namespaces this should look into ( `--background-color` and `--color` in that order).

Reply:
@thecrypticace Yeah :/ Kinda like this right? https://github.com/tailwindlabs/tailwindcss/pull/14177/files#diff-3e0c118787a4d539fc94ae61398cd2b7117ce62372be52991f873946336d3aa6R214-R241

Reply:
We (read: you) figured a way around this — thanks!